Fix rag finetuning + add finetuning test

lhoestq opened the pull request:
Following #7715 we need more test coverage of the RAG example scripts.
In this PR I'm adding a test for the finetuning script.
The test includes a single gpu test and a multi gpu test. Both are passing.

As mentioned in #7816 and #8345 there were some errors in the script that I had to fix.

Moreover since @amogkam has been working on the finetuning script as well to integrate Ray, I made sure to reduce the possible conflicts with his PR #8583 . More precisely I'm reusing the CustomAccel class that will allow to init either the pytorch distributed retrieval or the ray distributed retrieval.

Also fixed a bug in RAG forward pass (see #8665 )

Fix #7816
Fix #8345

shamanez commented:
@lhoestq 

Hi, I tried to execute finetune.py on two GPUs.  It mainly fails with the following error. But when I run with the single GPU it works. I have also attached a screen shot.



**RuntimeError: [/pytorch/third_party/gloo/gloo/transport/tcp/pair.cc:575] Connection closed by peer [130.216.209.142]:55728**

![Selection_006](https://user-images.githubusercontent.com/16892570/99660519-57831900-2ac7-11eb-933b-a99845ee3516.png)
![Selection_007](https://user-images.githubusercontent.com/16892570/99660537-5c47cd00-2ac7-11eb-9e91-272f2dd7f007.png)

lhoestq commented:
What command did you run exactly ?

shamanez commented:
> What command did you run exactly ?

`python examples/rag/finetune.py     --data_dir ./examples/rag/test_data/dummy_seq2seq     --output_dir ./examples/rag/outputs   --model_name_or_path facebook/rag-token-base      --model_type rag_sequence       --do_train     --do_predict     --n_val -1     --val_check_interval 0.25     --train_batch_size 1     --eval_batch_size 1     --max_source_length 128     --max_target_length 25     --val_max_target_length 25     --test_max_target_length 25     --label_smoothing 0.1     --dropout 0.1     --attention_dropout 0.1     --weight_decay 0.001     --adam_epsilon 1e-08     --max_grad_norm 0.1     --lr_scheduler polynomial     --learning_rate 3e-05     --num_train_epochs 100     --warmup_steps 500     --gradient_accumulation_steps 1    --index_name custom    --passages_path ./examples/rag/data/my_knowledge_dataset    --index_path  ./examples/rag/data/my_knowledge_dataset_hnsw_index.faiss --gpus 2
`

lhoestq commented:
Does changing the port with `--distributed_port 8888` help in your case ?

shamanez commented:
It says, 
`finetune.py: error: unrecognized arguments: --distributed_port 8888`

shamanez commented:
I tried with `--distributed-port 8888` still gives the same error. 
btw my torch version is **Version: 1.7.0+cu110**

lhoestq commented:
What's your pytorch lightning version ?
(also sorry I misspelled distributed-port)

shamanez commented:
> pytorch lightning

**Version: 1.0.4**

shamanez commented:
Hi just wanted to know .. did you managed to run the finetune.sh script
without any errors.

> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/huggingface/transformers/pull/8585#issuecomment-730323765>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEA4FGX4PJD6UFC5K24FA2TSQUBN7ANCNFSM4TYMOBIQ>
> .
>

lhoestq commented:
Yes I have no issue on my side. The finetuning test works fine too.
Could you try to update pytorch lightning and see if it fixes your issue ?
Let me know if you manage to fix it

shamanez commented:
Did you try with custom index ?

On Fri, Nov 20, 2020, 22:43 Quentin Lhoest <notifications@github.com> wrote:

> Yes I have no issue on my side. The finetuning test works fine too.
> Could you try to update pytorch lightning and see if it fixes your issue ?
> Let me know if you manage to fix it
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/huggingface/transformers/pull/8585#issuecomment-731061351>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEA4FGRRHXMLZB2KC5WTSJ3SQY25LANCNFSM4TYMOBIQ>
> .
>

lhoestq commented:
Let me test right now

shamanez commented:
Can you also send me your pytorch and tranformers versions.




On Fri, Nov 20, 2020, 22:49 Shamane Siriwardhana <shamane@ahlab.org> wrote:

> Did you try with custom index ?
>
> On Fri, Nov 20, 2020, 22:43 Quentin Lhoest <notifications@github.com>
> wrote:
>
>> Yes I have no issue on my side. The finetuning test works fine too.
>> Could you try to update pytorch lightning and see if it fixes your issue ?
>> Let me know if you manage to fix it
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/huggingface/transformers/pull/8585#issuecomment-731061351>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEA4FGRRHXMLZB2KC5WTSJ3SQY25LANCNFSM4TYMOBIQ>
>> .
>>
>

lhoestq commented:
Awesome I managed to reproduce your issue using the custom index :)
I will investigate
And I'm using pytorch 1.7.0 cu11 and transformers with the latest changes from master and this PR

shamanez commented:
Perfect. I feel it is some tensor issue that happens in the validation
sanity check.

On Fri, Nov 20, 2020, 23:02 Quentin Lhoest <notifications@github.com> wrote:

> Awesome I managed to reproduce your issue using the custom index :)
> I will investigate
> And I'm using pytorch 1.7.0 cu11 and transformers with the latest changes
> from master and this PR
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/huggingface/transformers/pull/8585#issuecomment-731071657>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEA4FGXTIZMVU76CGOR3FZTSQY5EHANCNFSM4TYMOBIQ>
> .
>

lhoestq commented:
Indeed it was an issue with the precision of the tensor. I'm fixing it

shamanez commented:
Amazing. If you can .. once you fixed this can you please add finetuning
commands for a custom dataset in the read me. The current one  is not with
all commands. I really think this RAG framework will be a game changer if
we can apply cor other tasks:)

On Fri, Nov 20, 2020, 23:18 Quentin Lhoest <notifications@github.com> wrote:

> Indeed it was an issue with the precision of the tensor. I'm fixing it
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/huggingface/transformers/pull/8585#issuecomment-731080487>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEA4FGWGLNSO3ITQGNLLUYTSQY7BFANCNFSM4TYMOBIQ>
> .
>

lhoestq commented:
Ok I fixed the tensor issue and updated the readme

I also had to rename some the examples files of RAG to avoid collisions with the files of the seq2seq examples. The name collision broke the CI tests with failed imports. 

I did:
```
examples/rag/utils.py -> exmaples/rag/utils_rag.py
examples/rag/callbacks.py -> exmaples/rag/callbacks_rag.py
examples/rag/finetune.py -> exmaples/rag/finetune_rag.py
examples/rag/finetune.sh -> exmaples/rag/finetune_rag.sh
```

All tests are green now :)

shamanez commented:
Thanks a lot for your quick response.

On Sat, Nov 21, 2020, 00:15 Quentin Lhoest <notifications@github.com> wrote:

> Ok I fixed the tensor issue and updated the readme
>
> I also had to rename some the examples files of RAG to avoid collisions
> with the files of the seq2seq examples. The name collision broke the CI
> tests with failed imports.
>
> I did:
>
> examples/rag/utils.py -> exmaples/rag/utils_rag.py
> examples/rag/callbacks.py -> exmaples/rag/callbacks_rag.py
> examples/rag/finetune.py -> exmaples/rag/finetune_rag.py
> examples/rag/finetune.sh -> exmaples/rag/finetune_rag.sh
>
> All tests are green now :)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/huggingface/transformers/pull/8585#issuecomment-731108211>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEA4FGTA626DKBJ5S7JXLFDSQZFVNANCNFSM4TYMOBIQ>
> .
>

patrickvonplaten commented on examples/rag/README.md:
```diff
@@ -130,3 +130,29 @@ python examples/rag/eval_rag.py \
+
+# Use your own knowledge source
+
+RAG uses by default the enghlish wikipedia as a knowledge source, known as the 'wiki_dpr' dataset.
```
```suggestion
By default, RAG uses the English Wikipedia as a knowledge source, known as the 'wiki_dpr' dataset.
```

patrickvonplaten commented on examples/rag/README.md:
```diff
@@ -130,3 +130,29 @@ python examples/rag/eval_rag.py \
+# Use your own knowledge source
+
+RAG uses by default the enghlish wikipedia as a knowledge source, known as the 'wiki_dpr' dataset.
+You can build your own knowledge source using the `use_custom_knowledge_dataset.py` and use it for RAG.
```
```suggestion
With `use_custom_knowledge_dataset.py` you can build your own knowledge source, *e.g.* for RAG.
```

patrickvonplaten commented on examples/rag/README.md:
```diff
@@ -130,3 +130,29 @@ python examples/rag/eval_rag.py \
+RAG uses by default the enghlish wikipedia as a knowledge source, known as the 'wiki_dpr' dataset.
+You can build your own knowledge source using the `use_custom_knowledge_dataset.py` and use it for RAG.
+
+Let's say you have documents in tab-separated csv files with columns "title" and "text", then you can run for example:
```
```suggestion
For instance, if documents are serialized as tab-separated csv files with the columns "title" and "text", one can use `use_own_knowledge_dataset.py` as follows:
```
Let's say you have documents in tab-separated csv files with columns "title" and "text", then you can run for example:
```

patrickvonplaten commented on examples/rag/README.md:
````diff
@@ -130,3 +130,29 @@ python examples/rag/eval_rag.py \
+    --output_dir path/to/my_knowledge_dataset \
+```
+
+To use it in the RAG finetuning script you can then do:
````
```suggestion
The created outputs in `path/to/my_knowledge_dataset` can then be used to finetune RAG as follows:
```

patrickvonplaten commented on examples/rag/callbacks_rag.py:
```diff
@@ -38,7 +38,7 @@ def get_checkpoint_callback(output_dir, metric):
         mode="max",
         save_top_k=3,
-        period=0,  # maybe save a checkpoint every time val is run, not just end of epoch.
+        period=1,  # maybe save a checkpoint every time val is run, not just end of epoch.
```
why go from 0 to 1?

lhoestq replied:
Oh I changed that to speed the the test and forgot to remove it. I can just modify the validation frequency

patrickvonplaten commented on examples/rag/finetune_rag.py:
```diff
@@ -429,10 +442,24 @@ def add_retriever_specific_args(parser):
+            "--use_dummy_dataset",
+            type=bool,
+            default=False,
+            help="If True, use the dummy version of the dataset index. More info about custom indexes in the RagRetriever documentation as well as in `examples/rag/use_own_knowledge_dataset.py`",
```
```suggestion
            help="Whether to use the dummy version of the dataset index. More info about custom indexes in the RagRetriever documentation as well as in `examples/rag/use_own_knowledge_dataset.py`",
```

patrickvonplaten replied:
This is usually the description for boolean args in Transformers I think -> see training_args.py

patrickvonplaten commented on examples/rag/test_data/dummy_seq2seq/test.source:
```diff
@@ -0,0 +1,2 @@
+What is love ?
```
Do you want to leave those in master?

patrickvonplaten replied:
If it's just for testing I'd rather create them at test time

lhoestq replied:
What do you recommend as an alternative ?
I can generate the dummy data on the fly in the test script for example

patrickvonplaten commented on src/transformers/models/rag/modeling_rag.py:
```diff
@@ -556,7 +556,9 @@ def forward(
             if has_to_retrieve:
-                question_enc_outputs = self.question_encoder(input_ids, attention_mask=attention_mask)
+                question_enc_outputs = self.question_encoder(
+                    input_ids, attention_mask=attention_mask, return_dict=True
```
Do we have to do this ? I know that modeling_rag.py is currently not compatible with tuples, but it would actually be better to implement this functionality with tuples rather than forcing the ModelOutput....

lhoestq replied:
Do you think that `question_enc_outputs = self.question_encoder(...)[0]` would be better then ? (taking the embeddingss from the tuple)

patrickvonplaten replied:
Actually, thinking more about since it's not possible to use tuples anyways, doing `return_dict=True` is fine (I do it as well in `generate()`....). Think we can just wait until someone posts an issue about not being able to use tuples with RAG. All good as it is for me - Sorry!

lhoestq replied:
I'm pretty sure we can use tuples actually. Why wouldn't it work ?

patrickvonplaten replied:
I would work! It'll just be a horrible `if-else cascade` to know which position to take for which cases. If you look at T5 you can see that there are already quite a lot of ugly if else statements to retrieve the correct position. In Rag it'll be much worse

patrickvonplaten reviewed (commented):
Awesome test! Just two things if we could I'd rather avoid them:
1. Don't like adding dummy data files to master
2. Would be nice to not force `return_dict=True`

lhoestq commented:
I took your comment into account @patrickvonplaten 
The only thing I didn't change is the return_dict=True - I kept them to avoid playing with tuples indices.

marcoaleixo commented:
@lhoestq hello, thank you for this amazing feature.

when I try to create my custom dataset I receveing this error:

`2020-12-16 00:48:44.645715: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
INFO:__main__:Step 1 - Create the dataset
Using custom data configuration default
Reusing dataset csv (/root/.cache/huggingface/datasets/csv/default-d44cf86c96b535d8/0.0.0/2960f95a26e85d40ca41a230ac88787f715ee3003edaacb8b1f0891e9f04dda2)
Loading cached processed dataset at /root/.cache/huggingface/datasets/csv/default-d44cf86c96b535d8/0.0.0/2960f95a26e85d40ca41a230ac88787f715ee3003edaacb8b1f0891e9f04dda2/cache-ad363af188e673b0.arrow
100% 1/1 [00:00<00:00, 10.92ba/s]
INFO:__main__:Step 2 - Index the dataset
Traceback (most recent call last):
  File "examples/rag/use_own_knowledge_dataset.py", line 200, in <module>
    main(rag_example_args, processing_args, index_hnsw_args)
  File "examples/rag/use_own_knowledge_dataset.py", line 102, in main
    index = faiss.IndexHNSWFlat(index_hnsw_args.d, index_hnsw_args.m, faiss.METRIC_INNER_PRODUCT)
  File "/usr/local/lib/python3.6/dist-packages/faiss/swigfaiss.py", line 3746, in __init__
    this = _swigfaiss.new_IndexHNSWFlat(*args)
NotImplementedError: Wrong number or type of arguments for overloaded function 'new_IndexHNSWFlat'.
  Possible C/C++ prototypes are:
    faiss::IndexHNSWFlat::IndexHNSWFlat()
    faiss::IndexHNSWFlat::IndexHNSWFlat(int,int)`


I'm using Google Colab to test this - https://colab.research.google.com/drive/1Cjj18rYmeS0Bueis_KPB5Wbybl-JNDLL?usp=sharing

marcoaleixo commented:
Well, i didn't install the specific dependencies you defined. excuse me.

Solved running - !pip install -r /transformers/examples/rag/requirements.txt

At least it is registered if someone has the same problem. haha
